src/client/src/producer.rs

```diff
@@ -216,6 +225,44 @@ fn assemble_requests(
     requests
 }
 
+/// A key for determining which partition a record should be sent to.
+///
+/// This type is used to support conversions from any other type that
+/// may be converted to a `Vec<u8>`, while still allowing the ability
+/// to explicitly state that a record may have no key (`RecordKey::Null`).
+///
+/// # Examples
+///
+/// ```
+/// # use fluvio::{TopicProducer, FluvioError, RecordKey};
+/// # async fn example(producer: &TopicProducer) -> Result<(), FluvioError> {
+/// producer.send("Hello", String::from("World!")).await?;
+/// producer.send(RecordKey::Null, "World!").await?;
+/// # Ok(())
+/// # }
+/// ```
+pub enum RecordKey {
+    Null,
+    Key(Vec<u8>),
```

Comment:
This should be `Bytes` to make it easier to translate to Record Type

Reply:
I have looked at how to go about this in the code and I don't think it's the right call. The fact that we use `Bytes` is a detail that is only evident down at the request level, an implementation detail that should not matter to the user. Also, coverting `Vec<u8>` to `Bytes` is cheap, it just takes ownership of the backing buffer.

Reply:
Implementation matter because you want to reduce unnecessary layer of abstraction.   Since we are using `Bytes` as internal representation we should propagate much as possible

Reply:
A more concrete reason to use `Vec<u8>` is that since `send` calls `send_all` and `send_all` accepts `Into<Vec<u8>>`, that means that we get all sorts of problems if we use `Bytes` in `send`:

```rust
    pub async fn send<K, V>(&self, key: K, value: V) -> Result<(), FluvioError>
    where
        K: Into<RecordKey>,
        V: Into<RecordValue>,
    {
        let record_key = key.into();
        let record_value = value.into();

        let maybe_key = match record_key { // This would become Option<Bytes>
            RecordKey::Null => None,
            RecordKey::Key(key) => Some(key),
        };
        let value = record_value.0; // This becomes `Bytes`

        // the trait `From<bytes::Bytes>` is not implemented for `Vec<u8>`
        self.send_all(Some((maybe_key, value))).await?;
        Ok(())
    }
```